### ISSUE TITLE
Best/recommended method for hyperparameter tuning

### ISSUE THREAD

**Issue by offendo** [facebookresearch/fairseq#3982]
#### What is your question?
Hello everyone,

I am using fairseq to implement & train a custom model. What's the best way to do a hyperparameter search (random search, grid search, etc.) within fairseq? I was hoping not to have to do it by hand :smile:  

In case it matters, here's what I'm doing now: 
1. I implemented a model in `my_model.py`
2. Put it into the `fairseq/models` folder
3. Run `fairseq-train` with CLI flags

This works fine for training. 

#### What have you tried?
Nothing, yet. My first thought was that it's possible to hook into the `Trainer` class training loop and add some code there, but I'm also not really sure how I would do that. Open to any and all suggestions!

#### What's your environment?

 - fairseq Version: `main`
 - PyTorch Version: 1.9.1
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): from source
 - Build command you used (if compiling from source): `python setup.py build_ext --inplace`
 - Python version: 3.9.7
 - CUDA/cuDNN version: 11.2
 - GPU models and configuration: GeForce RTX 3090

Thanks in advance!

**Comment by ashokrajab** [comment]
Hi @offendo,
I'm also trying to perform hyperparameter tuning for the model that I trained using fairseq. If you had figured out an efficient way, it would be helpful if you could share the same.

Thank you.

**Comment by offendo** [comment]
Hi @ashokrajab, I wish I had a better suggestion, but I ended up doing it manually which was rather tedious and probably not ideal...but it worked well enough for my use. My method looked like this

1. Calculate grid of hyperparameters (e.g., learning rate log-uniform between 1e-2 and 1e-5, dropout uniform between 0.1 and 0.5, etc.)
2. Train each configuration for a small number of epochs (I just wrote a bash script to launch one after another)
3. Log the performance at the end of each run, use the best configuration.
